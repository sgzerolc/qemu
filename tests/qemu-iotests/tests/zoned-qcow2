#!/usr/bin/env bash
#
# Test zone management operations for qcow2 file.
#

seq="$(basename $0)"
echo "QA output created by $seq"
status=1 # failure is the default!

file_name="zbc.qcow2"
_cleanup()
{
  _cleanup_test_img
  _rm_test_img "$file_name"
}
trap "_cleanup; exit \$status" 0 1 2 3 15

# get standard environment, filters and checks
. ../common.rc
. ../common.filter
. ../common.qemu

# This test only runs on Linux hosts with qcow2 image files.
_supported_proto file
_supported_os Linux

echo
echo "=== Initial image setup ==="
echo

$QEMU_IMG create -f qcow2 $file_name -o size=768M -o zone_size=64M \
-o zone_capacity=64M -o zone_nr_conv=0 -o max_append_sectors=512 \
-o max_open_zones=0 -o max_active_zones=0 -o zoned_profile=zbc

IMG="--image-opts -n driver=qcow2,file.driver=file,file.filename=$file_name"
QEMU_IO_OPTIONS=$QEMU_IO_OPTIONS_NO_FMT

echo
echo "=== Testing a qcow2 img with zoned format ==="
echo
echo "case 1: if the operations work"

echo "(1) report the first zone:"
$QEMU_IO $IMG -c "zrp 0 1"
echo
echo "report the first 10 zones"
$QEMU_IO $IMG -c "zrp 0 10"
echo
echo "report the last zone:"
$QEMU_IO $IMG -c "zrp 0x2C000000 2" # 0x2C000000 / 512 = 0x160000
echo
echo
echo "(2) opening the first zone"
$QEMU_IO $IMG -c "zo 0 0x4000000" # 0x4000000 / 512 = 0x20000
echo "report after:"
$QEMU_IO $IMG -c "zrp 0 1"
echo
echo "opening the second zone"
$QEMU_IO $IMG -c "zo 0x4000000 0x4000000"
echo "report after:"
$QEMU_IO $IMG -c "zrp 0x4000000 1"
echo
echo "opening the last zone"
$QEMU_IO $IMG -c "zo 0x2C000000 0x4000000"
echo "report after:"
$QEMU_IO $IMG -c "zrp 0x2C000000 2"
echo
echo
echo "(3) closing the first zone"
$QEMU_IO $IMG -c "zc 0 0x4000000"
echo "report after:"
$QEMU_IO $IMG -c "zrp 0 1"
echo
echo "closing the last zone"
$QEMU_IO $IMG -c "zc 0x3e70000000 0x4000000"
echo "report after:"
$QEMU_IO $IMG -c "zrp 0x3e70000000 2"
echo
echo
echo "(4) finishing the second zone"
$QEMU_IO $IMG -c "zf 0x4000000 0x4000000"
echo "After finishing a zone:"
$QEMU_IO $IMG -c "zrp 0x4000000 1"
echo
echo
echo "(5) resetting the second zone"
$QEMU_IO $IMG -c "zrs 0x4000000 0x4000000"
echo "After resetting a zone:"
$QEMU_IO $IMG -c "zrp 0x4000000 1"
echo
echo
echo "(6) append write" # the physical block size of the device is 4096
$QEMU_IO $IMG -c "zrp 0 1"
$QEMU_IO $IMG -c "zap -p 0 0x1000 0x2000"
echo "After appending the first zone firstly:"
$QEMU_IO $IMG -c "zrp 0 1"
$QEMU_IO $IMG -c "zap -p 0 0x1000 0x2000"
echo "After appending the first zone secondly:"
$QEMU_IO $IMG -c "zrp 0 1"
$QEMU_IO $IMG -c "zap -p 0x4000000 0x1000 0x2000"
echo "After appending the second zone firstly:"
$QEMU_IO $IMG -c "zrp 0x4000000 1"
$QEMU_IO $IMG -c "zap -p 0x4000000 0x1000 0x2000"
echo "After appending the second zone secondly:"
$QEMU_IO $IMG -c "zrp 0x4000000 1"

# success, all done
echo "*** done"
rm -f $seq.full
status=0
