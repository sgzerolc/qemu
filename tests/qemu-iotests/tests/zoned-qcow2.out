QA output created by zoned-qcow2

=== Initial image setup ===

Formatting 'zbc.qcow2', fmt=qcow2 cluster_size=65536 extended_l2=off compression_type=zlib zoned_profile=zbc zone_size=67108864 zone_capacity=67108864 zone_nr_conv=0 max_append_sectors=512 max_active_zones=0 max_open_zones=0 size=805306368 lazy_refcounts=off refcount_bits=16

=== Testing a qcow2 img with zoned format ===

case 1: if the operations work
(1) report the first zone:
start: 0x0, len 0x20000, cap 0x20000, wptr 0x0, zcond:1, [type: 2]

report the first 10 zones
start: 0x0, len 0x20000, cap 0x20000, wptr 0x0, zcond:1, [type: 2]
start: 0x20000, len 0x20000, cap 0x20000, wptr 0x20000, zcond:1, [type: 2]
start: 0x40000, len 0x20000, cap 0x20000, wptr 0x40000, zcond:1, [type: 2]
start: 0x60000, len 0x20000, cap 0x20000, wptr 0x60000, zcond:1, [type: 2]
start: 0x80000, len 0x20000, cap 0x20000, wptr 0x80000, zcond:1, [type: 2]
start: 0xa0000, len 0x20000, cap 0x20000, wptr 0xa0000, zcond:1, [type: 2]
start: 0xc0000, len 0x20000, cap 0x20000, wptr 0xc0000, zcond:1, [type: 2]
start: 0xe0000, len 0x20000, cap 0x20000, wptr 0xe0000, zcond:1, [type: 2]
start: 0x100000, len 0x20000, cap 0x20000, wptr 0x100000, zcond:1, [type: 2]
start: 0x120000, len 0x20000, cap 0x20000, wptr 0x120000, zcond:1, [type: 2]

report the last zone:
start: 0x160000, len 0x20000, cap 0x20000, wptr 0x160000, zcond:1, [type: 2]


(2) opening the first zone
wps[0]: 0x0
report after:
start: 0x0, len 0x20000, cap 0x20000, wptr 0x0, zcond:3, [type: 2]

opening the second zone
wps[1]: 0x20000
report after:
start: 0x20000, len 0x20000, cap 0x20000, wptr 0x20000, zcond:3, [type: 2]

opening the last zone
wps[11]: 0x160000
report after:
start: 0x160000, len 0x20000, cap 0x20000, wptr 0x160000, zcond:3, [type: 2]


(3) closing the first zone
wps[0]: 0x0
report after:
start: 0x0, len 0x20000, cap 0x20000, wptr 0x0, zcond:4, [type: 2]

closing the last zone
zone close failed: Input/output error
report after:
start: 0x1f380000, len 0x20000, cap 0x20000, wptr 0x0, zcond:0, [type: 2]
start: 0x1f3a0000, len 0x20000, cap 0x20000, wptr 0x0, zcond:0, [type: 2]


(4) finishing the second zone
wps[1]: 0x40000
After finishing a zone:
start: 0x20000, len 0x20000, cap 0x20000, wptr 0x40000, zcond:14, [type: 2]


(5) resetting the second zone
wps[1]: 0x20000
After resetting a zone:
start: 0x20000, len 0x20000, cap 0x20000, wptr 0x20000, zcond:1, [type: 2]


(6) append write
start: 0x0, len 0x20000, cap 0x20000, wptr 0x0, zcond:4, [type: 2]
wps[0]: 0x18
After zap done, the append sector is 0x0
After appending the first zone firstly:
start: 0x0, len 0x20000, cap 0x20000, wptr 0x18, zcond:2, [type: 2]
wps[0]: 0x30
After zap done, the append sector is 0x18
After appending the first zone secondly:
start: 0x0, len 0x20000, cap 0x20000, wptr 0x30, zcond:2, [type: 2]
wps[1]: 0x20018
After zap done, the append sector is 0x20000
After appending the second zone firstly:
start: 0x20000, len 0x20000, cap 0x20000, wptr 0x20018, zcond:2, [type: 2]
wps[1]: 0x20030
After zap done, the append sector is 0x20018
After appending the second zone secondly:
start: 0x20000, len 0x20000, cap 0x20000, wptr 0x20030, zcond:2, [type: 2]
*** done
